File: frontend/src/app/services/studio.service.ts
import { Injectable } from '@angular/core';
import {StudioCreateDto, StudioDto, StudioForm} from "../dtos/studio";
import {Observable} from "rxjs";
import {
  ApproveStudioDto,
  CreatedStudioDto,
  StudioSearchDto,
  StudioInfoDto
} from "../dtos/studio";
import {HttpClient, HttpParams} from "@angular/common/http";
import {Faq} from "../dtos/faq";
import {Membership} from "../dtos/membership";
import {environment} from "../../environments/environment";
import {StudioUpdateDto} from "../dtos/studio-update-dto";
import {PaginatedResponse} from "../dtos/paginator";
import {Instructor, InstructorCreateDto} from "../dtos/instructor";

const baseUri = environment.backendUrl + '/api/v1/studios';

@Injectable({
  providedIn: 'root'
})
export class StudioService {

  constructor(private httpClient: HttpClient,) { }

  createStudio(studio: StudioCreateDto): Observable<CreatedStudioDto> {

    const formData = new FormData();

    for (const key in studio) {
      formData.append(key, studio[key]);
    }

    console.log(formData)
    return this.httpClient.post<CreatedStudioDto>(baseUri, formData);
  }

  getNotApprovedStudios(): Observable<StudioInfoDto[]> {
    return this.httpClient.get<StudioInfoDto[]>(baseUri + '/notApproved');
  }

  approveStudio(studio: ApproveStudioDto): Observable<StudioInfoDto> {
    return this.httpClient.put<StudioInfoDto>(baseUri + "/approve", studio);
  }

  updateStudioByStudioId(studio: StudioUpdateDto, studioId: number): Observable<StudioDto> {
    console.log(studio);
    console.log(baseUri + `/${studioId}`)
    return this.httpClient.put<StudioDto>(baseUri + `/${studioId}`, studio);
  }

  updateProfileImage(file: File, studioId: number): Observable<StudioDto> {
    const formData = new FormData()
    formData.append('file', file);

    return this.httpClient.put<StudioDto>(baseUri + `/${studioId}/update-profile-image`, formData);
  }


  deleteStudio(id: number): Observable<StudioForm> {
    return this.httpClient.delete<StudioForm>(baseUri + '/' + id);
  }

  getStudioById(id: number): Observable<StudioDto> {
    return this.httpClient.get<StudioDto>(baseUri + '/' + id);
  }

  addFAQ(id: number, faq: Faq): Observable<Faq> {
    return this.httpClient.post<Faq>(baseUri + '/' + id + '/faqs', faq);
  }

  addMembership(id: number, membership: Membership): Observable<Membership> {
    return this.httpClient.post<Membership>(baseUri + '/' + id + '/memberships', membership);
  }

  addFavouriteStudio(studioId: number): Observable<StudioDto> {
    return this.httpClient.post<StudioDto>(`${baseUri}/${studioId}/add-favourite`, null);
  }

  removeFavouriteStudio(studioId: number): Observable<StudioDto> {
    return this.httpClient.delete<StudioDto>(`${baseUri}/${studioId}/remove-favourite`);
  }

  getFavouriteStudios(): Observable<StudioDto[]> {
    console.log(baseUri + '/favourite')
    return this.httpClient.get<StudioDto[]>(baseUri + '/favourite');
  }

  searchStudios(studioSearch: StudioSearchDto): Observable<PaginatedResponse<StudioInfoDto>> {
    let params = new HttpParams();

    if (studioSearch.name) {
      params = params.append('name', studioSearch.name);
    }
    if (studioSearch.location) {
      params = params.append('location', studioSearch.location);
    }

    params = params.append('pageIndex', studioSearch.pageIndex.toString());
    params = params.append('pageSize', studioSearch.pageSize.toString());

    return this.httpClient.get<PaginatedResponse<StudioInfoDto>>(baseUri + '/search', { params });
  }

  getStudioIdByAdmin(email: string): Observable<number> {
    return this.httpClient.get<number>(baseUri + "/getStudioByAdmin/" + email);
  }

  isCurrentUserAdmin(studioId: number): Observable<boolean> {
    return this.httpClient.get<boolean>(baseUri + "/isAdmin/" + studioId);
  }

  addInstructor(instructor: InstructorCreateDto, studioId: number): Observable<Instructor> {
    const formData = new FormData();
    for (const key in instructor) {
      formData.append(key, instructor[key]);
    }

    return this.httpClient.post<Instructor>(baseUri + "/add-instructor/" + studioId, formData);
  }

  uploadGalleryImages(formData: FormData, studioId: number): Observable<StudioDto> {
    return this.httpClient.put<StudioDto>(`${baseUri}/${studioId}/add-gallery-images`, formData);
  }

}

File: backend/src/main/java/com/inso/sila/service/StudioService.java
package com.inso.sila.service;

import com.inso.sila.endpoint.dto.studio.studio.CreatedStudioDto;
import com.inso.sila.endpoint.dto.studio.faqs.FaqsDto;
import com.inso.sila.endpoint.dto.studio.instructor.InstructorCreateDto;
import com.inso.sila.endpoint.dto.studio.instructor.InstructorDto;
import com.inso.sila.endpoint.dto.studio.membership.MembershipDto;
import com.inso.sila.endpoint.dto.studio.studio.StudioApprovalDto;
import com.inso.sila.endpoint.dto.studio.studio.StudioCreateDto;
import com.inso.sila.endpoint.dto.studio.studio.StudioUpdateDto;
import com.inso.sila.endpoint.dto.studio.studio.StudioDto;
import com.inso.sila.endpoint.dto.studio.studio.StudioInfoDto;
import com.inso.sila.endpoint.dto.studio.studio.StudioSearchDto;
import com.inso.sila.exception.ConflictException;
import com.inso.sila.exception.NotFoundException;
import org.springframework.web.multipart.MultipartFile;
import com.inso.sila.exception.ValidationException;
import org.springframework.data.domain.Page;

import java.io.IOException;
import java.util.List;

/**
 * Service interface for managing studio-related operations.
 */
public interface StudioService {

    /**
     * Adds a new studio.
     *
     * @param studio the DTO containing studio details
     * @return the saved Studio entity
     */
    CreatedStudioDto addStudio(StudioCreateDto studio) throws IOException, ValidationException, ConflictException;

    /**
     * Updates an existing studio.
     *
     * @param studio the DTO containing updated studio details
     * @return the updated Studio entity
     */
    StudioInfoDto updateApprovalOfStudio(StudioApprovalDto studio) throws NotFoundException;


    /**
     * Updates an existing studio.
     *
     * @param studioUpdateDto the DTO containing updated studio details.
     * @param studioId the studio id.
     * @return the updated Studio entity.
     */
    StudioDto updateStudio(StudioUpdateDto studioUpdateDto, Long studioId) throws NotFoundException, SecurityException;


    /**
     * Updates the profile image of a studio by a given studio id.
     *
     * @param profileImage the new profile image.
     * @param studioId the studio id.
     * @return the updated studio dto.
     */
    StudioDto updateProfileImage(MultipartFile profileImage, Long studioId) throws IOException, ValidationException, NotFoundException, SecurityException;


    /**
     * Retrieves all studios.
     *
     * @return a list of all Studio entities
     */
    List<StudioDto> getAllStudios();

    /**
     * Retrieves all liked studios by user.
     *
     * @return a list of liked Studio entities for one User
     */
    List<StudioDto> getAllLikedStudiosForUser();

    /**
     * Retrieves all studios that are not approved.
     *
     * @return a list of Studio entities that are not approved
     */
    List<StudioInfoDto> getNotApprovedStudios();

    /**
     * Deletes a studio by its ID.
     *
     * @param studioId the ID of the studio to delete
     * @return the deleted Studio entity
     * @throws jakarta.persistence.EntityNotFoundException if the studio is not found
     */
    StudioDto deleteStudio(Long studioId) throws NotFoundException;

    /**
     * Get a studio by given ID.
     *
     * @param studioId the ID of the needed studio
     * @return the needed studio
     * @throws jakarta.persistence.EntityNotFoundException if the studio is not found
     */
    StudioDto getStudioById(Long studioId) throws NotFoundException;

    /**
     * Add a frequently asked question to the studio.
     *
     * @param faqs the FAQ that needs to be added
     * @param studioId the ID of the needed studio
     * @return the added FAQ
     * @throws jakarta.persistence.EntityNotFoundException if the studio is not found
     */
    FaqsDto addFaq(FaqsDto faqs, Long studioId) throws SecurityException, NotFoundException;

    /**
     * Creates a new membership and associates it with the specified studio.
     *
     * @param membership the DTO containing membership details (e.g., name, duration, price).
     * @param studioId      the ID of the studio to link this membership to.
     * @return the saved Membership entity.
     * @throws NotFoundException if the studio or related activity is not found.
     */
    MembershipDto addMembership(MembershipDto membership, Long studioId) throws NotFoundException, SecurityException;


    /**
     * Adds an existent studio to the favourite
     * studios list of the current application user.
     *
     * @param studioId the id of the studio, which is added to the favourites.
     * @return a dto of the favourite studio
     * @throws NotFoundException if the studio or the current user are not found.
     */
    StudioDto addFavouriteStudio(Long studioId) throws NotFoundException;

    /**
     * Removes an existent studio from the favourite
     * studios list of the current application user.
     *
     * @param studioId the id of the studio, which is removed from the favourites.
     * @return a dto of the studio
     * @throws NotFoundException if the studio or the current user are not found.
     */
    StudioDto removeFavouriteStudio(Long studioId) throws NotFoundException;

    Page<StudioInfoDto> searchByNameAndLocation(StudioSearchDto studioSearchDto);

    Long getStudioIdByAdmin(String email);

    Boolean isCurrentUserAdmin(Long studioId) throws NotFoundException;

    InstructorDto addInstructor(InstructorCreateDto instructor, Long studioId) throws IOException, NotFoundException, ValidationException, SecurityException;

    /**
     * Adds new galleryImages to an existing studio with a studio id.
     *
     * @param studioId the studio id.
     * @param files the images as multipart files.
     * @return the studio dto with the newly created gallery images.
     */
    StudioDto addGalleryImages(Long studioId, MultipartFile[] files) throws IOException, NotFoundException, SecurityException, ValidationException;
}
